### PR TITLE
AP_Frsky_Telem: added RPM support for SPort protocol 4 and 10

### PR DESCRIPTION
**Pull request by yaapu**
This adds RPM data for sensor instances 1 and 2 for SPort protocol 4,10 and for FPort/FPort2.
Tested on a matekf405-wing with a dual PWM input as RPM data.

Notes:
- Real FRSky rpm sensors support values in the range 1000-300000
- FRSky rpm sensor support negative values (ardupilot uses a float for rpm)
- rpm data for passthrough uses an int16_t in increments of 10 rpms, so passthrough_rpm=rpm*0.1, actual rpm range is -327680 to 327670

Note: my editor clears trailing spaces on save so there are a few ghost lines,if required I can revert those changes

### CODE REVIEW

**Comment by yaapu** [comment]
@peterbarker not sure how to test this, is there a way to have RPM info from SITL?

**Comment by peterbarker** [comment]
On Mon, 18 Jan 2021, Alessandro Apostoli wrote:

> @peterbarker not sure how to test this, is there a way to have RPM info from SITL?

A SITL Plane while flying should produce RPM measurements.

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -9062,20 +9100,21 @@ def test_frsky_sport(self):
...
-            0x10:  self.tfs_validate_baro_alt, # baro alt integer m
-            0x11:  self.tfs_validate_gps_speed, # gps speed integer m/s
-            0x14:  self.tfs_validate_yaw, # yaw in degrees
-            0x19:  lambda x : True, # gps speed decimal cm/s
-            0x21:  lambda x : True, # altitude decimal m
-            0x28:  self.tfs_validate_current1, # current A
-            0x30:  self.tfs_validate_vspeed, # vertical speed m/s
-            0x39:  self.tfs_validate_battery1, # battery 1 voltage
-            0x800: self.tf_validate_gps, # gps lat/lon
+            0x01:   self.tfs_validate_gps_alt, # gps altitude integer m
```
Comment (peterbarker) [inline Tools/autotest/common.py:9103]: Could have saved a lot of delta here by not having a leading zero in your new key :-)

**Review by peterbarker — APPROVED** [review]
Looks like it should work....

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
good to merge when CI passes

File: Tools/autotest/common.py (hunk 1 above)
Comment (yaapu) [reply Tools/autotest/common.py:9103]: haha, oops, I'll fix it :-)

File: libraries/AP_Frsky_Telem/AP_Frsky_Backend.h [hunk 2]
```diff
@@ -58,6 +58,7 @@ class AP_Frsky_Backend
 
     void calc_nav_alt(void);
     void calc_gps_position(void);
+    bool calc_rpm(const uint8_t instance, int32_t &value) const;
```
Comment (peterbarker) [inline libraries/AP_Frsky_Telem/AP_Frsky_Backend.h:61]: For this sort of thing it's worht adding either a `WARN_IF_UNUSED` or a note that the passed-in-parameter is not modified if the function returns false.

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp [hunk 3]
```diff
@@ -188,6 +190,16 @@ bool AP_Frsky_SPort_Passthrough::is_packet_ready(uint8_t idx, bool queue_empty)
     case BATT_2:
         packet_ready = AP::battery().num_instances() > 1;
         break;
+    case RPM:
+        {
```
Comment (peterbarker) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:194]: Usually we'd put the brace on the previous line.

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp [hunk 4]
```diff
@@ -188,6 +190,16 @@ bool AP_Frsky_SPort_Passthrough::is_packet_ready(uint8_t idx, bool queue_empty)
...
+            packet_ready = false;
+            const AP_RPM *rpm = AP::rpm();
+            if (rpm == nullptr) {
+                break;
+            }
+            packet_ready = rpm->num_sensors() > 0;
```
Comment (peterbarker) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:200]:
```suggestion
            packet_ready = AP::rpm() && APM::rpm()->num_sensors() > 0;
```

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.h [hunk 5]
```diff
@@ -67,9 +67,10 @@ class AP_Frsky_SPort_Passthrough : public AP_Frsky_SPort, public AP_RCTelemetry
         BATT_2 =        8,  // 0x5008 Battery 2 status
         BATT_1 =        9,  // 0x5008 Battery 1 status
         PARAM =         10, // 0x5007 parameters
-        UDATA =         11, // user data
+        RPM =           11, // 0x500A rpm sensors 1 and 2
+        UDATA =         12, // user data
 #if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
-        MAV =           12, // mavlite
+        MAV =           13,  // mavlite
```
Comment (peterbarker) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.h:73]: This enumeration isn't actually used now.  Should we remove it?

**Review by peterbarker — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp (hunk 4 above)
Comment (yaapu) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:200]: @peterbarker I'm sure you told me once to not rely on implicit cast and and favor check vs nullptr instead, which btw I found pretty consistent as a suggestion, I did see in the code using the somewhat simpler implicit cast quite ofter though

Comment (peterbarker) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:200]: That's true.  Sometimes I just get lazy.

Could also have been
```
            packet_ready = (AP::rpm() != nullptr) && APM::rpm()->num_sensors() > 0;
```

Not quite as pithy :-)

**Pull request merged**